File: netDxf/IO/DxfWriter.Views.cs
#region netDxf library licensed under the MIT License
//
//                       netDxf library
// Copyright (c) Daniel Carvajal (haplokuon@gmail.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
#endregion

using System;
using System.Diagnostics;
using netDxf.Header;
using netDxf.Tables;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private void WriteView(View view)
        {
            Debug.Assert(this.activeTable == DxfObjectCode.ViewTable);
            if (view.IsCameraPlottable && this.doc.DrawingVariables.AcadVer < DxfVersion.AutoCad2007)
            {
                throw new NotSupportedException("A plottable named-view camera requires AutoCAD 2007 DXF or later.");
            }
            this.chunk.Write(0, view.CodeName);
            this.chunk.Write(5, view.Handle);
            this.WriteDatabaseMetadata(view);
            this.chunk.Write(330, view.Owner.Handle);
            this.chunk.Write(100, SubclassMarker.TableRecord);
            this.chunk.Write(100, SubclassMarker.View);
            this.chunk.Write(2, this.EncodeNonAsciiCharacters(view.Name));
            this.chunk.Write(70, (short)view.Flags);
            this.chunk.Write(40, view.Height);
            this.chunk.Write(10, view.ViewCenter.X);
            this.chunk.Write(20, view.ViewCenter.Y);
            this.chunk.Write(41, view.Width);
            this.chunk.Write(11, view.ViewDirection.X);
            this.chunk.Write(21, view.ViewDirection.Y);
            this.chunk.Write(31, view.ViewDirection.Z);
            this.chunk.Write(12, view.Target.X);
            this.chunk.Write(22, view.Target.Y);
            this.chunk.Write(32, view.Target.Z);
            this.chunk.Write(42, view.LensLength);
            this.chunk.Write(43, view.FrontClippingPlane);
            this.chunk.Write(44, view.BackClippingPlane);
            this.chunk.Write(50, view.Rotation);
            this.chunk.Write(71, (short)view.ViewMode);
            this.chunk.Write(281, (short)view.RenderMode);
            this.WriteViewUcs(view.Ucs);
            if (this.doc.DrawingVariables.AcadVer >= DxfVersion.AutoCad2007)
            {
                this.chunk.Write(73, view.IsCameraPlottable ? (short)1 : (short)0);
            }
            this.WriteSunReference(view);
            this.WriteXData(view.XData);
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.Views.js
// Copyright (c) Daniel Carvajal and netDxf contributors. MIT License; see package LICENSE.
import { DxfVersion } from '../Header/DxfVersion.js';
import { SubclassMarker } from '../SubclassMarker.js';
import { EncodeDxfText } from '../../runtime/DxfStringEncoding.js';
import { WriteXData } from '../../runtime/DxfXDataIO.js';
import { WriteViewUcs } from './DxfViewUcs.js';
import { WriteDatabaseMetadata } from './DxfWriter.Objects.js';
import { WriteSunReference } from './DxfWriter.Sun.js';
import { NotSupportedException } from '../../runtime/Errors.js';
export function WriteView(chunk, document, view) {
            if (view.IsCameraPlottable && document.DrawingVariables.AcadVer < DxfVersion.AutoCad2007)
            {
                throw new NotSupportedException("A plottable named-view camera requires AutoCAD 2007 DXF or later.");
            }
            chunk.Write(0, view.CodeName);
            chunk.Write(5, view.Handle);
            WriteDatabaseMetadata(chunk, document, view);
            chunk.Write(330, view.Owner.Handle);
            chunk.Write(100, SubclassMarker.TableRecord);
            chunk.Write(100, SubclassMarker.View);
            chunk.Write(2, EncodeDxfText(view.Name, document.DrawingVariables.AcadVer));
            chunk.Write(70, view.Flags);
            chunk.Write(40, view.Height);
            chunk.Write(10, view.ViewCenter.X);
            chunk.Write(20, view.ViewCenter.Y);
            chunk.Write(41, view.Width);
            chunk.Write(11, view.ViewDirection.X);
            chunk.Write(21, view.ViewDirection.Y);
            chunk.Write(31, view.ViewDirection.Z);
            chunk.Write(12, view.Target.X);
            chunk.Write(22, view.Target.Y);
            chunk.Write(32, view.Target.Z);
            chunk.Write(42, view.LensLength);
            chunk.Write(43, view.FrontClippingPlane);
            chunk.Write(44, view.BackClippingPlane);
            chunk.Write(50, view.Rotation);
            chunk.Write(71, view.ViewMode);
            chunk.Write(281, view.RenderMode);
            WriteViewUcs(chunk, view.Ucs);
            if (document.DrawingVariables.AcadVer >= DxfVersion.AutoCad2007)
            {
                chunk.Write(73, view.IsCameraPlottable ? 1 : 0);
            }
            if (view.HasStoredLiveSection) chunk.Write(334, view.LiveSection === null ? "0" : view.LiveSection.Handle);
            WriteSunReference(chunk, document.DrawingVariables.AcadVer, view);
            WriteXData(chunk, () => document.DrawingVariables.AcadVer, view.XData);
}
